File: first/mergesort.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

#define max_size 1000000

void merge(int array[], int p, int q, int r) {

    int n1 = q - p + 1;
    int n2 = r - q;

    int L[n1], M[n2];

    for (int i = 0; i < n1; i++)
      L[i] = array[p + i];
    for (int j = 0; j < n2; j++)
      M[j] = array[q + 1 + j];

    // Maintain current index of sub-arrays and main array
    int i, j, k;
    i = j = 0;
    k = p;

    while (i < n1 && j < n2) {
      if (L[i] <= M[j]) {
        array[k] = L[i];
        i++;
      } else {
        array[k] = M[j];
        j++;
      }
      k++;
    }

    while (i < n1) {
      array[k] = L[i];
      i++;
      k++;
    }

    while (j < n2) {
      array[k] = M[j];
      j++;
      k++;
    }
}

// Divide the array into two subarrays, sort them and merge them
void merge_sort(int array[], int l, int r) {
  if (l < r) {

    // m is the point where the array is divided into two subarrays
    int m = l + (r - l) / 2;

    merge_sort(array, l, m);
    merge_sort(array, m + 1, r);

    merge(array, l, m, r);
  }
}


int main(int argc, char** argv){
    srand(time(0));
    int array[max_size];
    int size = atoi(argv[1]);
    if (size > max_size){
        printf("Size must be smaller than 100000000");
        exit(1);
    }
    
    for (int i=0; i < size; i++) {
        array[i] = i ;  // fill the array in order
    }
    for (int i=0; i < size; i++){
        int random = i + (rand() % (size - i));
        int temp = array[i]; 
        array[i] = array[random]; 
        array[random] = temp;
        
    } // shuffle, like card deck
    // who the fuck would want true random here?

    printf("Unsorted array:  ");
    for (int i=0; i < size; i++){
        if(i +1 < size) {
            printf("%i, ",array[i]);
        }
        else {
            printf("%i \n",array[i]);
        }
    }

    merge_sort(array, 0, size - 1);

    printf("Sorted array:    ");
    for (int i = 0; i < size ; i++){
        if(i +1 < size) {
            printf("%i, ",array[i]);
        } else {
            printf("%i \n",array[i]);
        }
    }
    return 0;
}

File: first/mergesort.py
import random, os

MAX_SIZE = 1000000

def merge_sort(array):
    if len(array) > 1:

        # r is the point where this thing is divide into 2 subarrays
        r = len(array)//2
        L = array[:r]
        M = array[r:]

        # Sort 2 halves
        merge_sort(L)
        merge_sort(M)

        i = j = k = 0

        while i < len(L) and j < len(M):
            if L[i] < M[j]:
                array[k] = L[i]
                i += 1
            else:
                array[k] = M[j]
                j += 1
            k += 1

        # When we run out of elements in either L or M,
        # pick up the remaining elements and put in A[p..r]
        while i < len(L):
            array[k] = L[i]
            i += 1
            k += 1

        while j < len(M):
            array[k] = M[j]
            j += 1
            k += 1


def random_number_array(size):
    random_gen = random.sample(range(0, size), size)
    return random_gen

def get_number_input():
    print("> Input array size: ")
    try:
        num = int(input())
        return num
    except Exception as m:
        print(f"{m}; Please input a real number.")
        return get_number_input()


if __name__ == "__main__":
    size = get_number_input()
    unsorted_array = random_number_array(size)
    print(f"Unsorted array  : {unsorted_array}")

    sorted_array = unsorted_array
    merge_sort(sorted_array)
    print(f"Sorted array    : {sorted_array}")